src/job.rs

```diff
@@ -253,285 +254,45 @@ impl Job {
-            queue.push_back(&job1);
-        }
-        assert_eq!(unsafe { queue.tail.as_ref().as_usize() }, 1);
-
-        unsafe {
-            queue.pop_back();
-        }
-        assert_eq!(queue.sentinel, queue.tail);
-    }
-
-    #[test]
-    fn push2_pop2_back() {
-        let mut queue = JobQueue::default();
-
-        assert_eq!(queue.sentinel, queue.tail);
-
-        let job1 = Job::from_usize(&1);
-        let job2 = Job::from_usize(&2);
-
-        unsafe {
-            queue.push_back(&job1);
-        }
-        assert_eq!(unsafe { queue.tail.as_ref().as_usize() }, 1);
-
-        unsafe {
-            queue.push_back(&job2);
-        }
-        assert_eq!(unsafe { queue.tail.as_ref().as_usize() }, 2);
-
-        unsafe {
-            queue.pop_back();
-        }
-        assert_eq!(unsafe { queue.tail.as_ref().as_usize() }, 1);
-
-        unsafe {
-            queue.pop_back();
-        }
-        assert_eq!(queue.sentinel, queue.tail);
-    }
-
-    #[test]
-    fn push_pop_front() {
-        let mut queue = JobQueue::default();
-
-        assert_eq!(queue.sentinel, queue.tail);
-
-        let job1 = Job::from_usize(&1);
-
-        unsafe {
-            queue.push_back(&job1);
-        }
-        assert_eq!(unsafe { queue.tail.as_ref().as_usize() }, 1);
-
-        let job = unsafe { queue.pop_front().unwrap() };
-        assert_eq!(job.as_usize(), 1);
-        assert!(job.prev.get().is_none());
-        assert!(job.fut_or_next.get().is_some());
-
-        unsafe {
-            job.drop();
-        }
-
-        assert_eq!(queue.sentinel, queue.tail);
-    }
-
-    #[test]
-    fn push2_pop2_front() {
-        let mut queue = JobQueue::default();
-
-        assert_eq!(queue.sentinel, queue.tail);
-
-        let job1 = Job::from_usize(&1);
-        let job2 = Job::from_usize(&2);
-
-        unsafe {
-            queue.push_back(&job1);
-        }
-        assert_eq!(unsafe { queue.tail.as_ref().as_usize() }, 1);
-
-        unsafe {
-            queue.push_back(&job2);
-        }
-        assert_eq!(unsafe { queue.tail.as_ref().as_usize() }, 2);
-
-        let job = unsafe { queue.pop_front().unwrap() };
-        assert_eq!(job.as_usize(), 1);
-        assert!(job.prev.get().is_none());
-        assert!(job.fut_or_next.get().is_some());
-
-        unsafe {
-            job.drop();
-        }
-
-        let job = unsafe { queue.pop_front().unwrap() };
-        assert_eq!(job.as_usize(), 2);
-        assert!(job.prev.get().is_none());
-        assert!(job.fut_or_next.get().is_some());
-
-        unsafe {
-            job.drop();
-        }
-
-        assert_eq!(queue.sentinel, queue.tail);
+    pub fn pop_front(&mut self) -> Option<Job> {
+        let val = self.0.pop_front();
+        let job = match val {
+            // SAFETY:
+            // `Job` is still alive as per contract in `push_back`
+            Some(j) => unsafe { j.as_ref() },
+            None => return None,
+        };
+        job.is_in_queue.set(false);
```

Comment:
```suggestion
        let job = self.0.pop_front()?;
```

You can just use the `?` operator to deal with the `None` case here.